### DISCUSSION TITLE
Report map and ATT (or GATT) data base for “BLEKeyboardPassword” project

### DISCUSSION THREAD

**Discussion by ytsnew** [earlephilhower/arduino-pico#2020] [Q&A]
I am analyzing “BLEKeyboardPassword”.

Q1. Which file does determine the HID report map originally?

Report map:
05 01 09 06 a1 01 85 ….. 00 co

Q2. The constants (_attdb_head[], _attdb_batt_hidhead and so on) are defined in “PicoBluetoothBLEHID.h”. Is there any manual explaining how to construct these constants?

Limitation of my analyzation:
The following numerical sequence is in the _attib_head[].

// 0x0001 PRIMARY_SERVICE-GAP_SERVICE
0x0a, 0x00, 0x02, 0x00, 0x01, 0x00, 0x00, 0x28, 0x00, 0x18,

0x0a, 0x00 : 0x000a the number of bytes in this sequence, I guess.
0x02, 0x00 : what does these two bytes mean?
0x01, 0x00 : 0x0001 handle number
0x00, 0x28 : 0x2800 type : (UUID) Primary Service
0x00, 0x18 : 0x1800 value : Generic Access service

yts

**Comment by earlephilhower** [comment]
That's a standard GATT descriptor.  It's taken from the BTStack library where they have a GATT compiler in Python and examples.  See https://github.com/bluekitchen/btstack/tree/72ef1732c954d938091467961e41f4aa9b976b34

**Comment by ytsnew** [comment]
Hi. Earlephilhower,

Thank you for your quick response.
It is very difficult for me to find the answers in the web site you show me.

I will be happy if you answer my questions within your project “arduino-pico”.

Q1. In which file do you define the report descriptor?
Q2. 0x02, 0x00 : what do these two bytes mean? These bytes seem not to appear in the Bluetooth data packets that I monitor.

yts
